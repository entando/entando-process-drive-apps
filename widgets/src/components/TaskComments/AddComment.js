import React from 'react';
import PropTypes from 'prop-types';
import i18next from 'i18next';
import withStyles from '@material-ui/core/styles/withStyles';
import { TextareaAutosize, Button, CircularProgress, Typography } from '@material-ui/core';

const MAX_CHARS = 200;

const styles = {
  addNoteField: {
    width: '100%',
    border: '1px solid #E5E6E7',
  },
  addNoteLabel: {
    color: '#A1A1A1',
    margin: '6px 2px',
  },
  addNoteLimit: {
    fontSize: '10px',
    margin: '2px',
  },
  actionButtons: {
    marginTop: '25px',
    display: 'flex',
    justifyContent: 'flex-end',
  },
};

class AddComment extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      comment: '',
    };

    this.onChangeComment = this.onChangeComment.bind(this);
    this.handleAddCommentClick = this.handleAddCommentClick.bind(this);
  }

  onChangeComment({ target: { value } }) {
    this.setState({ comment: value });
  }

  handleAddCommentClick() {
    const { comment } = this.state;
    if (comment.trim()) {
      const { onClickAddComment } = this.props;

      onClickAddComment(comment.trim());
    }
    this.setState({ comment: '' });
  }

  render() {
    const { comment } = this.state;
    const { classes, loading } = this.props;
    const remainingChars = MAX_CHARS - comment.length;

    return (
      <div>
        <label htmlFor="comment-entry" className={classes.addNoteLabel}>
          {i18next.t('taskComments.addNote')}
        </label>
        <TextareaAutosize
          id="comment-entry"
          className={classes.addNoteField}
          rowsMin={2}
          value={comment}
          onChange={this.onChangeComment}
          disabled={loading}
        />
        <Typography
          color={remainingChars < 0 ? 'error' : 'initial'}
          className={classes.addNoteLimit}
        >
          {i18next.t('taskComments.maxChars', { max: MAX_CHARS, rem: remainingChars })}
        </Typography>
        <div className={classes.actionButtons}>
          {loading ? (
            <Button disabled variant="outlined" color="primary">
              <CircularProgress size="20px" />
            </Button>
          ) : (
            <Button onClick={this.handleAddCommentClick} variant="outlined" color="primary">
              {i18next.t('common.add')}
            </Button>
          )}
        </div>
      </div>
    );
  }
}

AddComment.propTypes = {
  classes: PropTypes.shape({
    addNoteField: PropTypes.string,
    addNoteLabel: PropTypes.string,
    addNoteLimit: PropTypes.string,
    actionButtons: PropTypes.string,
  }).isRequired,
  onClickAddComment: PropTypes.func,
  loading: PropTypes.bool,
};

AddComment.defaultProps = {
  onClickAddComment: () => {},
  loading: false,
};

export default withStyles(styles)(AddComment);
