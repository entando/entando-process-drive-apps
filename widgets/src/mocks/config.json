{
  "payload": {
    "code": "pda_task_list",
    "config": {
      "groups": "[{\"label\":\"Administrators\",\"key\":\"Administrators\",\"checked\":true},{\"label\":\"broker\",\"key\":\"broker\",\"checked\":true},{\"label\":\"admin\",\"key\":\"admin\",\"checked\":false},{\"label\":\"rest-all\",\"key\":\"rest-all\",\"checked\":false},{\"label\":\"kiemgmt\",\"key\":\"kiemgmt\",\"checked\":true},{\"label\":\"kie-server\",\"key\":\"kie-server\",\"checked\":true},{\"label\":\"appraiser\",\"key\":\"appraiser\",\"checked\":false},{\"label\":\"manager\",\"key\":\"manager\",\"checked\":true},{\"label\":\"supplier\",\"key\":\"supplier\",\"checked\":true},{\"label\":\"IT\",\"key\":\"IT\",\"checked\":true},{\"label\":\"PM\",\"key\":\"PM\",\"checked\":true}]",
      "options": "[{\"key\":\"newPageOnClick\",\"label\":\"Open new page on table row click\",\"checked\":false},{\"key\":\"showClaim\",\"label\":\"Show Claim Button\",\"checked\":true},{\"key\":\"showComplete\",\"label\":\"Show Complete Button\",\"checked\":true}]",
      "columns": "[{\"name\":\"task-is-skipable\",\"position\":0,\"isVisible\":false},{\"name\":\"task-parent-id\",\"position\":1,\"isVisible\":false},{\"name\":\"task-created-by\",\"position\":2,\"isVisible\":true},{\"name\":\"task-input-data.application.property.saleprice\",\"position\":3,\"isVisible\":true},{\"name\":\"task-input-data.TaskName\",\"position\":4,\"isVisible\":true},{\"name\":\"initiator\",\"position\":5,\"isVisible\":true},{\"name\":\"task-created-on\",\"position\":6,\"isVisible\":true},{\"name\":\"task-priority\",\"position\":7,\"isVisible\":false},{\"name\":\"task-proc-def-id\",\"position\":8,\"isVisible\":true},{\"name\":\"task-form\",\"position\":9,\"isVisible\":true},{\"name\":\"task-skippable\",\"position\":10,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.name\",\"position\":11,\"isVisible\":true},{\"name\":\"task-actual-owner\",\"position\":12,\"isVisible\":true},{\"name\":\"task-activation-time\",\"position\":13,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.address\",\"position\":14,\"isVisible\":true},{\"name\":\"task-input-data.application.errors\",\"position\":15,\"isVisible\":true},{\"name\":\"task-name\",\"position\":16,\"isVisible\":true},{\"name\":\"task-input-data.application.property.address\",\"position\":17,\"isVisible\":true},{\"name\":\"task-subject\",\"position\":18,\"isVisible\":true},{\"name\":\"containerId\",\"position\":19,\"isVisible\":true},{\"name\":\"task-container-id\",\"position\":20,\"isVisible\":true},{\"name\":\"processDefinitionId\",\"position\":21,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.ssn\",\"position\":22,\"isVisible\":true},{\"name\":\"task-description\",\"position\":23,\"isVisible\":true},{\"name\":\"task-input-data.Skippable\",\"position\":24,\"isVisible\":true},{\"name\":\"task-workitem-id\",\"position\":25,\"isVisible\":true},{\"name\":\"task-expiration-time\",\"position\":26,\"isVisible\":true},{\"name\":\"task-input-data.application.property.age\",\"position\":27,\"isVisible\":true},{\"name\":\"task-status\",\"position\":28,\"isVisible\":true},{\"name\":\"task-input-data.application.mortgageamount\",\"position\":29,\"isVisible\":true},{\"name\":\"task-process-instance-id\",\"position\":30,\"isVisible\":true},{\"name\":\"task-input-data.ActorId\",\"position\":31,\"isVisible\":true},{\"name\":\"task-process-id\",\"position\":32,\"isVisible\":true},{\"name\":\"task-input-data.application.property.locale\",\"position\":33,\"isVisible\":true},{\"name\":\"task-input-data.application.amortization\",\"position\":34,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.creditrating\",\"position\":35,\"isVisible\":true},{\"name\":\"application\",\"position\":36,\"isVisible\":true},{\"name\":\"task-input-data.NodeName\",\"position\":37,\"isVisible\":true},{\"name\":\"task-proc-inst-id\",\"position\":38,\"isVisible\":true},{\"name\":\"task-input-data.application.downpayment\",\"position\":39,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.annualincome\",\"position\":40,\"isVisible\":true},{\"name\":\"task-id\",\"position\":41,\"isVisible\":true}]",
      "knowledgeSource": "kieStaging",
      "process": "itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
    }
  },
  "metaData": {
    "status": "draft"
  },
  "errors": []
}
