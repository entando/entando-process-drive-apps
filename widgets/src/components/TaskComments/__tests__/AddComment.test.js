import { render, fireEvent, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import AddComment from 'components/TaskComments/AddComment';

describe('<AddComment />', () => {
  it('renders snapshot correctly', async () => {
    const { container } = render(<AddComment />);

    expect(container).toMatchSnapshot();
  });

  it('does not call onClickAddComment on render', async () => {
    const onClickAddComment = jest.fn();
    render(<AddComment onClickAddComment={onClickAddComment} />);

    await wait(() => {
      expect(onClickAddComment).toHaveBeenCalledTimes(0);
    });
  });

  it('calls onClickAddComment when "Add" button is clicked', async () => {
    const onClickAddComment = jest.fn();
    const { findByLabelText, findByText } = render(
      <AddComment onClickAddComment={onClickAddComment} />
    );

    const nameField = await findByLabelText('taskComments.addNote');
    fireEvent.change(nameField, { target: { value: 'Comment text.' } });

    const addButton = await findByText('common.add');
    fireEvent.click(addButton);

    await wait(() => {
      expect(onClickAddComment).toHaveBeenCalledTimes(1);
      expect(nameField).toBeEmpty();
    });
  });

  it('does not call onClickAddComment when "Add" button is clicked without comment added', async () => {
    const onClickAddComment = jest.fn();
    const { findByText } = render(<AddComment onClickAddComment={onClickAddComment} />);

    const addButton = await findByText('common.add');
    fireEvent.click(addButton);

    await wait(() => {
      expect(onClickAddComment).toHaveBeenCalledTimes(0);
    });
  });

  it('does not show Add when loading', async () => {
    const { queryByText, container } = render(<AddComment loading />);

    await wait(() => {
      expect(container.querySelector('svg')).toHaveClass('MuiCircularProgress-svg');
      expect(queryByText('common.add')).not.toBeInTheDocument();
    });
  });
});
