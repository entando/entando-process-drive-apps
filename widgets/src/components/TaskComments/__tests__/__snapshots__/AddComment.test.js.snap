// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddComment /> renders snapshot correctly 1`] = `
<div>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root AddComment-addNoteField-1"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="standard-textarea"
        id="standard-textarea-label"
      >
        taskComments.addNote
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
          id="standard-textarea"
          rows="2"
        />
      </div>
    </div>
    <div
      class="AddComment-actionButtons-2"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          common.add
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
